### ISSUE TITLE
Possible false positive for `omit_obvious_property_types`

### ISSUE THREAD

**Issue by denniskaselow** [dart-lang/linter#5121]
This code:

```dart
final Map<int, String> foobar = {};
```

gets flagged as:
> The type annotation isn't needed because it is obvious.

I just want to make sure if that's expected behavior when the generics are not specified on the right side of the assignment.

**Comment by eernstg** [comment]
Right, if the type annotation is simply deleted then the type changes (and it's not a recommended typing because it has type arguments `<dynamic, dynamic>`). So we definitely don't want that.

However, the automatic fix operation that you can perform in some IDEs will not just remove the type annotation, it will add the type arguments to the map literal, yielding this:

```dart
final foobar = <int, String>{};
```

I think this style is quite useful (and it's been recommended by various members of the community): The type and value of the initializing expression is quite easy to see at a glance. With `final Map<int, String> foobar = {};` we'd have to look at the type annotation in order to see the type, and at the initializing expression in order to see the value.

Also, if you have several declarations of variables (all `final` or all `var`) whose type is "obvious" then you can have a nicely aligned column of names. So you might want to express the variable declarations such that they have an obvious type as often as possible.

Of course, you might prefer to have the type annotation more frequently (e.g., in _every_ variable declaration), but in that case you wouldn't enable `omit_obvious_property_types` in the first place.

So I think it's OK to lint this declaration.

However, I do agree that the lint message is a bit too terse in this case (it would have been nice if it could say that the actual type arguments should be moved from the left to the right, not just that the type annotation should be removed).

**Comment by denniskaselow** [comment]
I didn't see the quick fix. I usually use them via right click in the Dart Analysis view (IntelliJ) but there was none available:

![image](https://github.com/user-attachments/assets/2af31adc-475f-4536-876e-67cb6938f9fa)

It's available when I do it directly on the variable.

I guess the quick fix in the Dart Analysis view will become available with a new Dart Plugin version.

**Comment by eernstg** [comment]
It would be available in the lightbulb menu of VScode. I think IntelliJ is less well maintained these days, so that could be the reason, but I don't have a precise explanation why this happens.

**Comment by bwilkerson** [comment]
I also don't know the reason, but it isn't because IntelliJ is less well maintained. The support for fixes is shared between the major platforms. The fix ought to have shown up.

Would you mind opening a separate issue for the missing fix, preferably in [the SDK issue tracker](https://github.com/dart-lang/sdk/issues). Thanks!

**Comment by denniskaselow** [comment]
Opened a new issue here: https://github.com/dart-lang/sdk/issues/57017

**Comment by srawlins** [comment]
@eernstg do the docs help with the message? Because indeed for this code:

```dart
final Map<int, String> foobar = {};
```

The message:

> The type annotation isn't needed because it is obvious.

is false, right? The type annotation _does_ change the type. `{}` is not "obviously" a `Map<int, String>`; it is "obviously" a `Map<dynamic, dynamic>`. Do the linked docs clarify this messaging?

**Comment by eernstg** [comment]
> The message ... is false, right?

Ah, of course! The lint should detect that there are no syntactic type arguments on the map literal, and there are no key/value pairs that could determine those type arguments. It works for lists and sets, but, apparently, there is a bug in the handling of maps, so it emits the overly optimistic claim about removing the type annotation. Thanks, it's being fixed in https://dart-review.googlesource.com/c/sdk/+/393280!

**Comment by eernstg** [comment]
Fixed in c9391d7f19c19c012956eb74d00e0b0dcea82c41.
